controller/impl.go

```diff
@@ -21,44 +31,43 @@ func (f *Factory) startController(ctx context.Context, tenantID, policyID, polic
 		"host":           host,
 	}).Logger()
 
-	options, err := f.cfg.Server.ToConnectionOptions()
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to setup grpc dial options for the remote service")
-	}
-
-	options = append(options, client.WithTenantID(tenantID), client.WithDialOptions(f.dopts...))
+	errGroup := errgroup.Group{}
 
-	stop := make(chan bool)
-	cleanup := func() {
-		stop <- true
+	conn, err := f.cfg.Server.Connect(client.WithTenantID(tenantID), client.WithDialOptions(f.dopts...))
+	if err != nil {
+		return func() {}, errors.Wrap(err, "failed to initialize new connection")
 	}
 
-	go func() {
-		for {
-			err = f.runCommandLoop(ctx, &logger, policyID, policyName, instanceLabel, host, commandFunc, stop, options)
-			if err == nil || err == io.EOF {
-				return
+	remoteCli := management.NewControllerClient(conn)
+	ctx, cancel := context.WithCancel(ctx)
+
+	errGroup.Go(func() error {
+		for retry := 0; ; retry++ {
+			retry++
+			err = f.runCommandLoop(ctx, &logger, policyID, policyName, instanceLabel, host, commandFunc, remoteCli)
+			if err == nil { // graceful shutdown on context canceled.
+				return nil
 			}
-
 			logger.Info().Err(err).Msg("command loop exited with error, restarting")
-			time.Sleep(5 * time.Second)
+			backoff := timeout * time.Duration(math.Pow(2, float64(retry)))
+			if sleepWithContext(ctx, min(backoff, maxBackoff)) == Canceled {
+				break
+			}
 		}
-	}()
 
-	return cleanup, nil
-}
-
-func (f *Factory) runCommandLoop(ctx context.Context, logger *zerolog.Logger, policyID, policyName, instanceLabel, host string, commandFunc CommandFunc, stop <-chan bool, opts []client.ConnectionOption) error {
-	callCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+		return nil
+	})
 
-	conn, err := client.NewConnection(opts...)
-	if err != nil {
-		return errors.Wrap(err, "failed to connect to the control plane")
-	}
+	return func() {
+		cancel()
+		if err = errGroup.Wait(); err != nil {
```

Comment:
Don't reuse `err` from the outer scope. Not only is it confusing and error prone but in this case it also means that `err` would escape to the heap because it's being captured by the closure even though the outer function may have already returned.

```go
if err := errGroup.Wait(); err != nil {
    ...
}
```